src/utils/SocketClient.hxx

```diff
@@ -150,12 +183,23 @@ public:
     {
         {
             AtomicHolder h(this);
-            if (state_ != STATE_SHUTDOWN)
+            if (requestShutdown_)
             {
-                state_ = STATE_SHUTDOWN_REQUESTED;
+                return;
+            }
+            requestShutdown_ = true;
+            if (sleeping_)
+            {
+                sleeping_ = false;
+                timer_.ensure_triggered();
+            }
+            if (isConnected_)
+            {
+                isConnected_ = false;
+                notify();
             }
         }
-        sem_.post();
+        /// @todo wake up any pending tasks
```

Comment:
Does this todo comment need to be resolved before merging back?

Reply:
Today tasks are pending because we are in a blocking system call, such as in connect() or a getaddrinfo(). I don't have a way to wake those up so we'll have to just wait for them to finish and timeout.

Reply:
added comment to the code.